funnel/models/project.py

```diff
@@ -108,11 +107,42 @@ class Project(BaseScopedNameMixin, db.Model):
         'all': {
             'read': {
                 'id', 'name', 'title', 'datelocation', 'timezone', 'date', 'date_upto', 'url_json',
-                '_state', 'website', 'bg_image', 'bg_color', 'explore_url', 'tagline', 'url'
+                '_state', 'website', 'bg_image', 'bg_color', 'explore_url', 'tagline', 'url',
+                'location'
                 },
             },
         }
 
+    @cached_property
+    def datelocation(self):
+        """
+        Returns a date + location string for the event, the format depends on project dates
+
+        If it's a single day event
+        > 11 Feb 2018, Bangalore
+
+        If multi-day event in same month
+        > 09 - 12 Feb 2018, Bangalore
+
+        If multi-day event across months
+        > 27 Feb - 02 Mar 2018, Bangalore
+
+        If multi-day event across years
+        > 30 Dec 2018 - 02 Jan 2019, Bangalore
+        """
+        strfdate = "%d" if self.date.month == self.date_upto.month else "%d %b"  # format for <start date>
+        dateformat = "{date} - {date_upto} {year}, {location}"  # format for whole string
+
+        if self.date == self.date_upto:
+            # if both dates are same, in case of single day project
+            dateformat = "{date_upto} {year}, {location}"
+        elif self.date.year != self.date_upto.year:
+            # if the start date and end dates are in different years,
+            strfdate = "%d %b %Y"
```

Comment:
This logic is a bit convoluted since you start with defaults and then replace them. Write this as a series of if/elif conditions.

Reply:
Ideally we should be doing locale-specific renderings, but that has the potential to mix up DDMM vs MMDD based on misconfigured locales, so let's start here and revisit when necessary.